src/container/RelayRootContainer.js

```diff
@@ -141,6 +141,28 @@ class RelayRootContainer extends React.Component {
   _runQueries(
     {Component, forceFetch, refetchRoute, route}: RootContainerProps
   ): RootContainerState {
+    var fragmentNames = Component.getFragmentNames();
+    if (__DEV__) {
+      var unmatchedNameSet = {};
+      Object.keys(route.queries).forEach((name) => unmatchedNameSet[name] = 1);
+      fragmentNames.forEach(fragmentName => {
+        // XOR both sets to find unmatched names
+        if (unmatchedNameSet[fragmentName]) {
+          delete unmatchedNameSet[fragmentName];
+        } else {
+          unmatchedNameSet[fragmentName] = 1;
+        }
+      });
+      var unmatchedNames = Object.keys(unmatchedNameSet);
+      invariant(
```

Comment:
If we move this into `getRelayQueries` and can make this check cheaper, we can drop the `__DEV__` check. (In practice, there are not that many fragment names.)

Otherwise, we should convert this into a `warning`. Behavior of code should not be significantly altered by `__DEV__`, in order to guarantee that bugs that happen in production (where `__DEV__` is false) can be reproduced in development (where `__DEV__` is true). I'll add this to `CONTRIBUTING.md`.